contracts/vesting/src/vesting.rs

```diff
@@ -0,0 +1,894 @@
+                            .amount
+                            .u128(),
+                    ),
+            ),
+            denom: DENOM.to_string(),
+        })
+    }
+
+    fn spendable_coins(
+        &self,
+        block_time: Option<Timestamp>,
+        env: &Env,
+        storage: &dyn Storage,
+    ) -> Result<Coin, ContractError> {
+        Ok(Coin {
+            amount: Uint128::new(
+                self.load_balance(storage)?
+                    .u128()
+                    .saturating_sub(self.locked_coins(block_time, env, storage)?.amount.u128()),
+            ),
+            denom: DENOM.to_string(),
+        })
+    }
+
+    fn get_vested_coins(
+        &self,
+        block_time: Option<Timestamp>,
+        env: &Env,
+    ) -> Result<Coin, ContractError> {
+        let block_time = block_time.unwrap_or(env.block.time);
+        let period = self.get_current_vesting_period(block_time);
+
+        let amount = match period {
+            // We're in the first period, or the vesting has not started yet.
+            0 => Coin {
+                amount: Uint128::new(0),
+                denom: DENOM.to_string(),
+            },
+            // We always have 8 vesting periods, so periods 1-7 are special
+            1..=7 => Coin {
+                amount: Uint128::new(self.tokens_per_period()? * period as u128),
+                denom: DENOM.to_string(),
+            },
+            _ => Coin {
+                amount: self.coin.amount,
+                denom: DENOM.to_string(),
+            },
+        };
+        Ok(amount)
+    }
+
+    fn get_vesting_coins(
+        &self,
+        block_time: Option<Timestamp>,
+        env: &Env,
+    ) -> Result<Coin, ContractError> {
+        Ok(Coin {
+            amount: Uint128::new(
+                self.get_original_vesting().amount.u128()
+                    - self.get_vested_coins(block_time, env)?.amount.u128(),
+            ),
+            denom: DENOM.to_string(),
+        })
+    }
+
+    fn get_start_time(&self) -> Timestamp {
+        self.start_time
+    }
+
+    fn get_end_time(&self) -> Timestamp {
+        self.periods[(NUM_VESTING_PERIODS - 1) as usize].end_time()
+    }
+
+    fn get_original_vesting(&self) -> Coin {
+        self.coin.clone()
+    }
+
+    fn get_delegated_free(
+        &self,
+        block_time: Option<Timestamp>,
+        env: &Env,
+        storage: &dyn Storage,
+    ) -> Result<Coin, ContractError> {
+        let block_time = block_time.unwrap_or(env.block.time);
+        let period = self.get_current_vesting_period(block_time);
+        let max_vested = self.tokens_per_period()? * period as u128;
+        let start_time = self.periods[period].start_time;
+
+        let delegations_keys = self
+            .delegations()
+            .keys_de(storage, None, None, Order::Ascending)
+            .scan((), |_, x| x.ok())
+            .filter(|(_mix, block_time)| block_time < &start_time)
```

Comment:
Instead of adding a reference and letting the compiler auto-dereference, you could directly dereference it yourself.
```suggestion
            .filter(|(_mix, block_time)| *block_time < start_time)
```